### DISCUSSION TITLE
Model Server Theia Integration : Changes not reflected in the coffee.ecore from the api-test-menu option "edit"

### DISCUSSION THREAD

**Discussion by NomsaT** [eclipse-emfcloud/emfcloud#120] [ModelServer] [answered]
From the api-test-menu, the edit api request does not update the changes as it should(in the coffee.ecore model) from the code. By that I mean  for  example looking at the following code:
 
       commands.registerCommand(EditSetCoffeeEcoreCommand, {
            execute: () => {
                const owner = {
                    'eClass':
                        'http://www.eclipse.org/emf/2002/Ecore#//EClass',
                    '$ref':
                        `${this.workspaceUri}/Coffee.ecore#//@eClassifiers.2`
                };
                const feature = 'name';
                const changedValues = ['ControlUnitNew'];
                const setCommand = new SetCommand(owner, feature, changedValues);
                this.modelServerClient
                    .edit('Coffee.ecore', setCommand)
                    .then(result => this.printResponse('edit SetCommand', result));
            }
        });

I would expect the changes to be reflected after saving the coffee.ecore model by clicking the save(coffee.ecore) from the api-test-menu, but the initial and the "updated" model are the same. Is there anything or steps we are missing here or our understanding of this command/code is distorted?

**Comment by ndoschek** [comment]
Hi @NomsaT!
Your assumptions are correct, after saving your `Coffee.ecore` resource file should be updated.
I tested the described workflow on current master and cannot reproduce the behaviour, i.e. the name of the third `eClassifier` was successfully changed to `ControlUnitNew` after saving the resource.
Could you please provide more details to repdroduce? Thanks!

**Reply by NomsaT** [reply]
Hi @ndoschek :)

Thank you so much for the reply.

I am not sure if you meant to say you were able to reproduce the behavior  "I tested the described workflow on current master and cannot reproduce the behavior" but essentially the flow is make changes to the coffee.ecore file using the edit :(edit(Coffee.ecore,{type:set})) option from the api-test-menu and saving the model after using the save: save(Coffee.ecore) from the api-test-menu options as well. So after this I would expect the name of the third eClassifier as you have mentioned in your comment above  to change from "ControlUnit" to "ControlUnitNew" in the Coffee.ecore resource file but none of that is happening, the initial and the "updated"  files  are still the same

**Reply by ndoschek** [reply]
Hi @NomsaT!

Sorry that I did not make myself clear enough.
With current master, everything works as expected for me. After saving the resource file is updated as expected.

I recorded a short GIF:
![emfcloud-discussions-120](https://user-images.githubusercontent.com/47414588/155164323-0ef72c48-efce-4385-89a2-b65e465a7aee.gif)

Do you get the correct notifications (also as shown in the GIF) after triggering the commands via the menu? How do you check the files? Do you get any error messages?

**Comment by NomsaT** [comment]
Hi @ndoschek 

Thank you so much for the effort.

This doesn't seem to work on my side and I am not sure where the problem is. I do get the notifications after triggering the commands via the menu. I did exactly what you did in terms of checking the file. I get no error messages and in the model server it does shows that the "put" API request has been received. 

Funny enough if I try to undo before the changes I get a notification that says that "cannot undo" which is correct but if I try to undo after the "saved" changes  I also get notification that says "successfully undo" which is correct I guess assuming that changes were actually made, but the name of the third eClassifier is and was still the same.

surely there's something I am missing.

**Reply by ndoschek** [reply]
Hi @NomsaT,
... if I try to undo after the "saved" changes I also get notification that says "successfully undo" which is correct I guess ...

To me that sounds that the actual Model Server version you are using is outdated. ~~With the current Model Server version an undo request after saving is not possible anymore.~~ Edit: sorry that was a wrong info.

- Do you start the Model Server separately from the [`emfcloud-modelserver` repository](https://github.com/eclipse-emfcloud/emfcloud-modelserver) or do you start only the `emfcloud-modelserver-theia` backend with the Model Server jar?
- Did you make changes to the Model Server?

**Comment by NomsaT** [comment]
HI @ndoschek 

>Actaully used both options for testing purposes. Just discovered that when using saveAll from the api-test-menu options after editing the Coffee.ecore the changes are finally reflected, but the save doesn't seem to work. Btw this worked with the internal server.
>No I didn't make any changes to the model server.

**Reply by ndoschek** [reply]
Hi @NomsaT,
Which OS do you use? I tested with Linux Ubuntu - it could be a path problem perhaps?

**Comment by NomsaT** [comment]
Hi @ndoschek 

I am using Windows thought so too initially but it doesn't seem like it.
I will test it on Linux as well and give you some feedback. 

Thank you so much for your help

**Comment by NomsaT** [comment]
Hi @ndoschek 

Quick update: We have tested the repo on Linux and it works perfectly fine.
We also discovered that yes It is  a path issue. 

https://github.com/cloudsmith/org.eclipse.emf/blob/0426d3bd36353798a212cb859ea4b5fbef301826/plugins/org.eclipse.emf.common/src/org/eclipse/emf/common/util/URI.java#L216

Otherwise thank you so much for your help :)

**Comment by ndoschek** [comment, marked as answer]
Hi @NomsaT,
Thank you for the update! Glad it works now on Linux!

I checked the behaviour in meantime on a Windows VM and I could reproduce the same issues there as well.
I opened [a bug on the emfcloud-modelserver repository](https://github.com/eclipse-emfcloud/emfcloud-modelserver/issues/173) that addresses that issue.

**Comment by NomsaT** [comment]
Hi @ndoschek 

Thank you so much. Will look out for the updates.

Enjoy the rest of your day :)
